packages/flutter/lib/src/widgets/shortcuts.dart

```diff
@@ -140,24 +139,145 @@ class KeySet<T extends KeyboardKey> {
+///    incoming event in asertion order.
+///  * Of the previous list, finds the first activator whose [accepts] returns
+///    true.
+///
+/// See also:
+///
+///  * [SingleActivator], an implementation that represents a single key combined
+///    with modifiers (control, shift, alt, meta).
+///  * [LogicalKeySet], an implementation that requires one or more
+///    [LogicalKeyboardKey]s to be pressed at the same time. Prefer
+///    [SingleActivator] when possible.
+abstract class ShortcutActivator {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const ShortcutActivator();
+
+  /// All the keys that might be the final event to trigger this shortcut.
+  ///
+  /// For example, for `Ctrl-A`, the KeyA is the only trigger, while Ctrl is not,
+  /// because the shortcut should only work by pressing KeyA *after* Ctrl, but
+  /// not before. For `Ctrl-A-E`, on the other hand, both KeyA and KeyE should be
+  /// triggers, since either of them is allowed to trigger.
+  ///
+  /// The trigger keys are used as the first-pass filter for incoming events, as
+  /// [Intent]s are stored in a [Map] and indexed by trigger keys. Subclasses
+  /// should make sure that the return value of this method does not change
+  /// throughout the lifespan of this object.
+  Iterable<LogicalKeyboardKey>? get triggers;
+
+  /// Whether the triggering `event` and the keyboard `state` at the time of the
+  /// event meet required conditions, providing that the event is a triggering
+  /// event.
+  ///
+  /// For example, for `Ctrl-A`, it has to check if the event is a
+  /// [RawKeyDownEvent], if either side of the Ctrl key is pressed, and none of
+  /// the Shift keys, Alt keys, or Meta keys are pressed; it doesn't have to
+  /// check if KeyA is pressed, since it's already guaranteed.
+  ///
+  /// This method must not cause any side effect to `state`. Typically
+  /// this is only used to query whether [RawKeyboard.keysPressed] contains
+  /// a key.
+  ///
+  /// See also:
+  ///
+  ///  * [LogicalKeyboardKey.collapseSynonyms], which helps deciding whether a
+  ///    modifier key is pressed when the side variation is not important.
+  bool accepts(RawKeyEvent event, RawKeyboard state);
+
+  /// Returns a description of the key set that is short and readable.
+  ///
+  /// Intended to be used in debug mode for logging purposes.
+  String debugDescribeKeys();
+}
+
 /// A set of [LogicalKeyboardKey]s that can be used as the keys in a map.
 ///
-/// A key set contains the keys that are down simultaneously to represent a
-/// shortcut.
+/// [LogicalKeySet] can be used as a [ShortcutActivator]. It is not recommended
+/// to use [LogicalKeySet] for a common shortcut such as `Delete` or `Ctrl+C`,
+/// prefer [SingleActivator] when possible, whose behavior more closely resembles
+/// that of typical platforms.
```

Comment:
That was my previous implementation, but then I decided against it.

`LogicalKeySet` currently produces a weird behavior (as described here). There are two possible fates `LogicalKeySet`:

*  **(The current choice)** Implement `LogicalKeySet` in `ShortcutActivator` while keeping its behavior exactly as is. In the future we will deprecate `LogicalKeySet`, and possibly create a new class that has a behavior similar to `LogicalKeySet` but much more regular (assume it's called `KeySetActivator`).
* Change `LogicalKeySet` so that it behaves in the same way as the current `SingleActivator`. 

The previous implementation chose the 2nd way, but resulted in some weird behavior due to backward compatibility, such as the question raised in https://github.com/flutter/flutter/pull/78522#discussion_r597836007.

**More detailed explanation to the question for the previous implementation:** Basically it judges whether a `LogicalKeySet` was "regular" by checking if it contains any non-modifier keys (so that non-modifier keys can be used as `triggers`). For "regular" key combinations `LogicalKeySet` should ignore sides because we're removing synonym modifier keys (no more `alt`, only `altLeft` and `altRight`). For "irregular" key combinations `LogicalKeySet` should consider sides because `KeySetActivator` is supposed to.

Reply:
OK, makes sense.